core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -80,8 +87,58 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
   //periodically emit metrics (don't need to do this for each message!)
   context.system.scheduler.schedule(30.seconds, 10.seconds, self, EmitMetrics)
 
+  // Key is ColdStartKey, value is the number of cold Start in minute
+  var coldStartCount = immutable.Map.empty[ColdStartKey, Int]
+
   backfillPrewarms(true)
 
+  // check periodically every 1 minute, adjust prewarmed container(delete if unused for some time and create some increment containers)
+  context.system.scheduler.schedule(1.minute, 1.minute, self, AdjustPrewarmedContainer)
+
+  def adjustPrewarmedContainer(): Unit = {
+    prewarmConfig.foreach { config =>
+      // Delete unused prewarmed container until minCount is reached
+      val kind = config.exec.kind
+      val memory = config.memoryLimit
+      val ttlSeconds = config.ttl.toSeconds
+      val minCount = config.minCount
+      val containers = prewarmedPool.filter { warmInfo =>
+        warmInfo match {
+          case (_, PreWarmedData(_, `kind`, `memory`, _)) => true
+          case _                                          => false
+        }
+      }
+      var currentCount = containers.size
+      for ((container, data) <- containers) {
+        if (currentCount > minCount && JDuration
+              .between(data.lastUsed, Instant.now)
+              .compareTo(JDuration.ofSeconds(ttlSeconds)) > 0) {
+          container ! RemovePreWarmedContainer
+          currentCount -= 1
+        }
+      }
+
+      // supplement some prewarmed container if cold start happened
+      val calendar = Calendar.getInstance()
+      calendar.add(Calendar.MINUTE, -1)
+      val lastDate = new SimpleDateFormat("yyyy-MM-dd-HH:mm").format(calendar.getTime)
+      val coldStartKey = ColdStartKey(kind, memory.toMB, lastDate)
+      coldStartCount.get(coldStartKey) match {
+        case Some(value) =>
+          if (value >= config.threshold) {
+            val createdCount = value / config.increment
+            val count = if (createdCount > 0) createdCount else 1
+            prewarmContainerIfpossible(kind, memory, count)
+          } else {
+            // at lease create 1 prewarmed container
+            prewarmContainerIfpossible(kind, memory, 1)
+          }
+        case None =>
+      }
+      coldStartCount = coldStartCount - coldStartKey
```

Comment:
Should this just reset the map to empty? Otherwise this map may grow without bound?

Reply:
Yes, reset the map's key of kind/memory to empty.

Reply:
I mean that it might be good to simple reset the map at once, instead of removing each key separately.

Reply:
Updated accordingly